lib/preview/previewUtils.js

```diff
@@ -0,0 +1,183 @@
+      if (timeWaited !== nextWait) return;
+      timeWaited = 0;
+      const res = fetch(window.location.origin + '/meta')
+        .then(async (res) => {
+          const hsServerState = await res.json();
+          if (hsServerState["REMOTE_FS_IS_DIRTY"]) {
+            location.reload();
+          }
+          nextWait = NORMAL_WAIT_MS;
+        })
+        .catch(err => {
+          nextWait *= BACKOFF_RATIO;
+          console.log('Disconnected from local server... (retrying in ' + nextWait / 1000 + 's)');
+        });
+    }, NORMAL_WAIT_MS);
+  })();
+  </script>
+  `;
+  return insertAtEndOfBody(html, refreshScript);
+}
+
+const getSubDomainFromValidLocalDomain = hostname => {
+  for (const validProxyDomainSuffix of VALID_PROXY_DOMAIN_SUFFIXES) {
+    if (hostname.endsWith(`.${validProxyDomainSuffix}`)) {
+      return hostname.slice(0, -1 * validProxyDomainSuffix.length - 1);
+    }
+  }
+  return undefined;
+};
+
+// From(ish) https://git.hubteam.com/HubSpot/cloudflare-workers/blob/master/worker-lib/src/Constants.ts
+const internalRoutes = {
+  HCMS: '/_hcms/',
+  HS_FS: '/hs-fs/',
+  HUB_FS: '/hubfs/'
+}
+
+const isInternalCMSRoute = (req) =>
+  Object.values(internalRoutes).some((route => req.path.startsWith(route)));
+
+const silenceConsoleWhile = async (act, ...args) => {
+  const tmpConsole = console;
+  console = { log: () => {} }
+  const result = await act(...args);
+  console = tmpConsole;
+  return result;
+}
+
+const memoize = (func, cacheBustCallback) => {
+  const cache = {};
+  return async (...args) => {
+    const stringArgs = args.toString();
+    const storedResult = cache[stringArgs];
+    if (storedResult && (cacheBustCallback ? !cacheBustCallback() : true)) {
+      //console.log(`Cache hit ${func}`)
+      return storedResult;
+    }
+    //console.log(`Cache miss ${func}`)
+    const res = await func(...args);
+    cache[stringArgs] = res;
+    return res;
+  }
+}
+
+const hidePreviewInDest = (previewDest) => previewDest.split('/').slice(2).join('/');
+
+const trackPreviewEvent = async (action) => {
+  if (!isTrackingAllowed()) {
+    return;
+  }
+  const accountId = getAccountId();
+
+  trackUsage(
+    'cli-interaction',
+    'INTERACTION',
+    {
+      applicationName: 'hubspot.preview',
+      os: `${platform()} ${release()}`,
+      authType: getAuthType(accountId),
+      action,
+    },
+    accountId
+  ).catch(
+    (err) => {
+      console.error(`trackUsage failed: ${JSON.stringify(err, null, 2)}`);
+    }
+  );
+}
+
+const getAuthType = (accountId) => {
+  let authType = 'unknown';
+
+  if (accountId) {
+    const accountConfig = getAccountConfig(accountId);
+    authType =
+      accountConfig && accountConfig.authType
+        ? accountConfig.authType
+        : 'apikey';
+  }
+
+  return authType;
+};
+
+const isUngatedForPreview = async (sessionInfo) => {
+  const { portalId, env, personalAccessKey } = sessionInfo;
+
+  const { enabledFeatures = {} } = await unAuth.fetchAccessToken(
```

Comment:
`fetchAccessToken ` is not necessarily safe to call because it's possible that the access token has expired. I think you'll need to call `accessTokenForPersonalAccessKey` or maybe break up the logic in there to determine whether or not the access token needs to be refreshed.